### DISCUSSION TITLE
Page Caching Mechanism

### DISCUSSION THREAD

**Discussion by mdmontesinos** [oqtane/oqtane.framework#4486] [Q&A]
Does Oqtane have any page caching system already in place?

If so, how does it work and how can we configure it?

For example, I would like to, not only cache a oqtane "page" with all its submodules (that should be invalidated when any of them is updated) but the whole HTML output of the requested route. In this way, I could also cache the theme as it requires some database queries for dynamic menus.

**Comment by sbwalker** [comment]
@mdmontesinos I assume you are referring to "page output caching" ie. where the rendered version of a page is cached on the server or client (or both) based on a caching strategy. And I assume you are only referring to the scenario where Oqtane is using Static Rendering and all of the modules on a page are statically rendered (meaning that the user is either not logged in or does not have the right to edit any content on the page). As it is only under this specific use case where "page output caching" could potentially be used - although I am not event sure if Blazor's Enhanced Navigation may still cause a conflict (as it is not doing a full page refresh on every request).

**Reply by mdmontesinos** [reply]
@sbwalker Apologies, I should have been more specific. Yes, I'm refering to page output caching in static render mode and no interactive components. In a regular Blazor application, I could use the OutputCache attribute on the components with the Page directive to cache the whole page output (see https://github.com/dotnet/aspnetcore/issues/49130). In the case of Oqtane, is there any way to achieve this type of caching?

**Reply by sbwalker** [reply]
@mdmontesinos currently this capability is not available in Oqtane. An option could be added in Page Management however there are a lot of considerations based on the fact that Oqtane composes the UI dynamically at run-time based on configuration:

ie. if page caching is enabled for a page

- it should be ignored if the Site is not using Static rendering
- it should be ignored if the user is logged in and has Edit rights to the page
- it should be ignored if any component on the page is being rendered interactively

Some of the validation logic above is complicated and has the potential to negatively impact performance. This will require some deeper investigation.

**Reply by mdmontesinos** [reply]
@sbwalker You're right that it has a lot of considerations and has the potential to break so many things. Is there any other way currently to at least cache all the page modules? Otherwise I'll just include a memory cache in the most time consuming ones.

**Reply by sbwalker** [reply]
@mdmontesinos Oqtane already includes caching for a variety of scenarios. When a user navigates to a page, all of the configuration for that page (ie. page settings, modules, permissions, etc...) are all retrieved from the cache (rather than calling the database). In addition, the HtmlText module stores its content in the cache. You could certainly cache the content of your own modules as well. Note that this is not "output caching" as it caches the data rather than the actual HTML output.

**Reply by mdmontesinos** [reply]
@sbwalker Ok, as I mentioned, I will resort to manually caching the content for the most resource consuming modules. Anyway, I believe it would be interesting to have a page output caching in Oqtane, as it is extremelly useful for sites with thousands of users. In my use case, the vast majority of users (not logged in) will arrive at the home page and I need it to be served as instantly as possible.
